src/controllers/bulk/bulkenumerator.cpp

```diff
@@ -2,60 +2,73 @@
 
 #include <libusb.h>
 
+#include <QDebug>
+#include <algorithm>
+#include <memory>
+#include <span>
+
 #include "controllers/bulk/bulkcontroller.h"
 #include "controllers/bulk/bulksupported.h"
 #include "moc_bulkenumerator.cpp"
+#include "util/assert.h"
 
-BulkEnumerator::BulkEnumerator(UserSettingsPointer pConfig)
+BulkEnumerator::BulkEnumerator()
         : ControllerEnumerator(),
-          m_context(nullptr),
-          m_pConfig(pConfig) {
-    libusb_init(&m_context);
+          m_context(nullptr, &libusb_exit) {
+    // TODO: use C++23 std::out_ptr instead
+    libusb_context* pCtx;
+    libusb_init(&pCtx);
+    m_context.reset(pCtx);
 }
 
 BulkEnumerator::~BulkEnumerator() {
     qDebug() << "Deleting USB Bulk devices...";
-    while (m_devices.size() > 0) {
-        delete m_devices.takeLast();
-    }
-    libusb_exit(m_context);
 }
 
-static bool is_interesting(struct libusb_device_descriptor *desc) {
-    for (int i = 0; bulk_supported[i].vendor_id; ++i) {
-        if ((bulk_supported[i].vendor_id == desc->idVendor) &&
-            (bulk_supported[i].product_id == desc->idProduct)) {
-            return true;
-        }
-    }
-    return false;
+static bool is_interesting(const libusb_device_descriptor& desc) {
+    auto vendorId = desc.idVendor;
+    auto productId = desc.idProduct;
+    return std::any_of(std::cbegin(bulk_supported),
+            std::cend(bulk_supported),
+            [=](const auto& dev) {
+                return dev.vendor_id == vendorId && dev.product_id == productId;
+            });
 }
 
 QList<Controller*> BulkEnumerator::queryDevices() {
     qDebug() << "Scanning USB Bulk devices:";
-    libusb_device **list;
-    ssize_t cnt = libusb_get_device_list(m_context, &list);
-    ssize_t i = 0;
-    int err = 0;
-
-    for (i = 0; i < cnt; i++) {
-        libusb_device *device = list[i];
+    libusb_device** deviceListPointers; // contiguous array of pointers to libusb_device
+    ssize_t cnt = libusb_get_device_list(m_context.get(), &deviceListPointers);
+    VERIFY_OR_DEBUG_ASSERT(cnt >= 0) {
+        // this assumes that if libusb_get_device_list returned an error, no memory was allocated
+        // and listPtr is still uninitialized. Unfortunately the libusb docs don't specify the state
+        // of the arguments in the case of errors.
+        qWarning() << "Error retrieving devicelist: " << libusb_error_name(cnt);
+    }
+    auto pDeviceList = std::unique_ptr<libusb_device*[],
+            decltype([](libusb_device** list) {
+                libusb_free_device_list(list, true);
+            })>(deviceListPointers);
```

Comment:
This looks bulky. How about creating `deviceListDeleter()` in the anonymous namespace?